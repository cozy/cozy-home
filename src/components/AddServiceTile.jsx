import React from 'react'
import { useClient, generateWebLink } from 'cozy-client'
import AppLinker from 'cozy-ui/transpiled/react/AppLinker'

import SquareAppIcon from 'cozy-ui/transpiled/react/SquareAppIcon'

const AddServiceTile = ({ label }) => {
  const client = useClient()
  const nativePath = '/discover/?type=konnector'
  const slug = 'store'
  const cozyURL = new URL(client.getStackClient().uri)
  const { subdomain: subDomainType } = client.getInstanceOptions()

  return (
    <AppLinker
      app={{ slug: 'store' }}
      nativePath={nativePath}
      href={generateWebLink({
        pathname: '/',
        cozyUrl: cozyURL.origin,
        slug,
        hash: nativePath,
        subDomainType
      })}
    >
      {({ onClick, href }) => (
        <a onClick={onClick} href={href} className="scale-hover">
          <SquareAppIcon name={label} variant="add" />
        </a>
      )}
    </AppLinker>
  )
}

export default AddServiceTile
