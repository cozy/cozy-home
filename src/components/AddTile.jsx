import React, { useState } from 'react'

import flag from 'cozy-flags'
import { useClient, generateWebLink } from 'cozy-client'
import AppLinker from 'cozy-ui/transpiled/react/AppLinker'
import SquareAppIcon from 'cozy-ui/transpiled/react/SquareAppIcon'
import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'

import styles from '@/styles/lists.styl'
import ShortcutCreateModal from './Shortcuts/ShortcutCreateModal'
/**
 * AddTile component.
 *
 * @returns {JSX.Element} The rendered AddTile component.
 */
const AddTile = ({ isStoreAvailable }) => {
  const client = useClient()
  const nativePath = '/discover'
  const slug = 'store'
  const cozyURL = new URL(client.getStackClient().uri)
  const { subdomain: subDomainType } = client.getInstanceOptions()
  const { t } = useI18n()
  const [isCreateModalOpen, setIsCreateModalOpen] = useState(false)

  if (flag('home.add-tile.add-shortcut')) {
    return (
      <>
        <a
          onClick={() => setIsCreateModalOpen(true)}
          className={styles['scale-hover']}
        >
          <SquareAppIcon name={t('add_service')} variant="add" />
        </a>
        {isCreateModalOpen && (
          <ShortcutCreateModal onClose={() => setIsCreateModalOpen(false)} />
        )}
      </>
    )
  }

  if (isStoreAvailable) {
    return (
      <AppLinker
        app={{ slug: 'store' }}
        nativePath={nativePath}
        href={generateWebLink({
          pathname: '/',
          cozyUrl: cozyURL.origin,
          slug,
          hash: nativePath,
          subDomainType
        })}
      >
        {({ onClick, href }) => (
          <a onClick={onClick} href={href} className={styles['scale-hover']}>
            <SquareAppIcon name={t('add_service')} variant="add" />
          </a>
        )}
      </AppLinker>
    )
  }

  return null
}

export default AddTile
