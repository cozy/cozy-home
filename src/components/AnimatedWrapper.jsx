import React from 'react'
import { isFlagshipApp } from 'cozy-device-helper'
import homeConfig from 'config/home.json'
import { useOpenApp } from 'hooks/useOpenApp'
import { RemoveScroll } from 'react-remove-scroll'

import App from 'containers/App'

const RemoveScrollOnFlaghsip = ({ children }) => {
  if (isFlagshipApp()) {
    return <RemoveScroll forwardProps>{children}</RemoveScroll>
  }

  return children
}

const AnimatedWrapper = () => {
  const { getAppState } = useOpenApp()

  const mainStyle = isFlagshipApp()
    ? {
        position: 'fixed',
        height: '100%'
      }
    : {}

  return (
    <RemoveScrollOnFlaghsip>
      <div
        className={`App ${getAppState} u-flex u-flex-column u-w-100 u-miw-100 u-flex-items-center`}
        style={mainStyle}
      >
        <App {...homeConfig} />
      </div>
    </RemoveScrollOnFlaghsip>
  )
}

export default AnimatedWrapper
