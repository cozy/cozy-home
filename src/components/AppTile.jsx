import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { models } from 'cozy-client'

import { translate } from 'cozy-ui/transpiled/react/I18n'
import AppLinker from 'cozy-ui/transpiled/react/AppLinker'
import SquareAppIcon from 'cozy-ui/transpiled/react/SquareAppIcon'
const { applications } = models

export class AppTile extends Component {
  render() {
    const { app, lang } = this.props
    const displayName = applications.getAppDisplayName(app, lang)
    const appHref = app.links && app.links.related
    return (
      <AppLinker slug={app.slug} href={appHref} app={app}>
        {({ onClick, href }) => (
          <a onClick={onClick} href={href} className="scale-hover">
            <SquareAppIcon app={app} name={displayName} />
          </a>
        )}
      </AppLinker>
    )
  }
}

AppTile.propTypes = {
  app: PropTypes.object.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(AppTile)
