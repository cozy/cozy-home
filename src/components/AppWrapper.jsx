/* global __DEVELOPMENT__ */

import React, { createContext } from 'react'
import { Provider as ReduxProvider } from 'react-redux'
import memoize from 'lodash/memoize'

import flag from 'cozy-flags'
import CozyClient, { CozyProvider } from 'cozy-client'
import CozyDevtools from 'cozy-client/dist/devtools'
import I18n from 'cozy-ui/transpiled/react/I18n'
import MuiCozyTheme from 'cozy-ui/transpiled/react/MuiCozyTheme'
import { BreakpointsProvider } from 'cozy-ui/transpiled/react/hooks/useBreakpoints'

import { CozyProvider as LegacyCozyProvider } from 'lib/redux-cozy-client'
import configureStore from 'store/configureStore'
import homeConfig from 'config/home.json'
import { CozyClient as LegacyCozyClient } from 'lib/redux-cozy-client'
import { RealtimePlugin } from 'cozy-realtime'

import schema from '../schema'

const dictRequire = lang => require(`locales/${lang}.json`)

export const AppContext = createContext()

/**
 * Setups clients and store
 *
 * Is memoized to avoid several clients in case of hot-reload
 */
export const setupAppContext = memoize(() => {
  const lang = document.documentElement.getAttribute('lang') || 'en'
  const context = window.context || 'cozy'
  const root = document.querySelector('[role=application]')
  const data = root.dataset

  // New improvements must be done with CozyClient
  const cozyClient = new CozyClient({
    uri: `${window.location.protocol}//${data.cozyDomain}`,
    schema,
    token: data.cozyToken
  })

  cozyClient.registerPlugin(flag.plugin)
  cozyClient.registerPlugin(RealtimePlugin)

  const legacyClient = new LegacyCozyClient({
    cozyURL: `//${data.cozyDomain}`,
    token: data.cozyToken,
    cozyClient
  })

  // store
  const store = configureStore(legacyClient, cozyClient, context, {
    lang,
    ...homeConfig
  })

  return { cozyClient, store, data, lang, context }
})

/**
 * Setups the app context and creates all context providers and wrappers
 * for an app
 */
const AppWrapper = ({ children }) => {
  const appContext = setupAppContext()
  const { store, cozyClient, data, context, lang } = appContext
  return (
    <AppContext.Provider value={appContext}>
      <BreakpointsProvider>
        <MuiCozyTheme>
          <CozyProvider client={cozyClient}>
            <LegacyCozyProvider
              store={store}
              client={cozyClient}
              domain={data.cozyDomain}
              secure={!__DEVELOPMENT__}
            >
              <ReduxProvider store={store}>
                <I18n lang={lang} dictRequire={dictRequire} context={context}>
                  {children}
                  {process.env.NODE_ENV !== 'production' ? (
                    <CozyDevtools />
                  ) : null}
                </I18n>
              </ReduxProvider>
            </LegacyCozyProvider>
          </CozyProvider>
        </MuiCozyTheme>
      </BreakpointsProvider>
    </AppContext.Provider>
  )
}

export default AppWrapper
