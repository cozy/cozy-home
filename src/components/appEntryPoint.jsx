import CozyClient, { Q, queryConnect } from 'cozy-client'

const ACCOUNT_DOCTYPE = 'io.cozy.accounts'
const KONNECTOR_DOCTYPE = 'io.cozy.konnectors'
const TRIGGER_DOCTYPE = 'io.cozy.triggers'

const OLDER_THAN_THIRTY_SECONDS = CozyClient.fetchPolicies.olderThan(30 * 1000)

const appEntryPoint = queryConnect({
  accounts: {
    query: () => Q(ACCOUNT_DOCTYPE),
    as: 'accounts',
    fetchPolicy: OLDER_THAN_THIRTY_SECONDS
  },
  konnectors: {
    query: () => Q(KONNECTOR_DOCTYPE),
    as: 'konnectors',
    fetchPolicy: OLDER_THAN_THIRTY_SECONDS
  },
  triggers: {
    query: () => Q(TRIGGER_DOCTYPE),
    as: 'triggers',
    fetchPolicy: OLDER_THAN_THIRTY_SECONDS
  }
})

export default appEntryPoint
