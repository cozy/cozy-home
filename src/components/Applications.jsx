import React, { memo, useEffect, useRef } from 'react'
import memoize from 'lodash/memoize'

import { useQuery } from 'cozy-client'
import flag from 'cozy-flags'
import Divider from 'cozy-ui/transpiled/react/Divider'
import { useI18n } from 'cozy-ui/transpiled/react'

import AppTile from 'components/AppTile'
import LogoutTile from 'components/LogoutTile'
import ShortcutLink from 'components/ShortcutLink'
import LoadingPlaceholder from 'components/LoadingPlaceholder'
import homeConfig from 'config/home.json'
import useHomeShortcuts from 'hooks/useHomeShortcuts'
import { appsConn } from 'queries'

import SquareAppIcon from 'cozy-ui/transpiled/react/SquareAppIcon'

const LoadingAppTiles = memo(({ num }) => {
  const { t } = useI18n()
  return (
    <>
      {Array(num)
        .fill(null)
        .map((e, i) => (
          <div key={i}>
            <SquareAppIcon
              variant="ghost"
              name={t('loading.working')}
              IconContent={<LoadingPlaceholder />}
            />
          </div>
        ))}
    </>
  )
})
LoadingAppTiles.displayName = LoadingAppTiles

const isValidData = memoize(data => Array.isArray(data) && data.length > 0)

const getApplicationsList = memoize(data =>
  isValidData(data) ? (
    data
      .filter(
        app =>
          app.state !== 'hidden' &&
          !homeConfig.filteredApps.includes(app.slug) &&
          !flag(`home_hidden_apps.${app.slug.toLowerCase()}`) // can be set in the context with `home_hidden_apps: - drive - banks`for example
      )
      .map(app => <AppTile key={app.id} app={app} />)
  ) : (
    <LoadingAppTiles num={3} />
  )
)

export const Applications = ({ onAppsFetched }) => {
  const showLogout = !!flag('home.mainlist.show-logout')
  const shortcuts = useHomeShortcuts()
  const { data } = useQuery(appsConn.query, appsConn)
  const didLoad = useRef(false)

  useEffect(() => {
    const isReady =
      didLoad.current === false && onAppsFetched && isValidData(data)

    isReady && onAppsFetched(data) && (didLoad.current = true)
  }, [data, onAppsFetched])

  return (
    <div className="app-list-wrapper u-m-auto u-w-100">
      <Divider className="u-mv-0" />

      <div className="app-list u-w-100 u-mv-3 u-mv-2-t u-mh-auto u-flex-justify-center">
        {getApplicationsList(data)}

        {shortcuts.map((shortcut, index) => (
          <ShortcutLink key={index} file={shortcut} />
        ))}

        {showLogout && <LogoutTile />}
      </div>
    </div>
  )
}

export default Applications
