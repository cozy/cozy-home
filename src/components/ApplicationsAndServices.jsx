import React from 'react'

import flag from 'cozy-flags'

import LogoutTile from 'components/LogoutTile'
import ShortcutLink from 'components/ShortcutLink'

import { useApps } from './Applications'
import { useServices } from './Services'
import AddTile from './AddTile'

export const ApplicationsAndServices = () => {
  const showLogout = !!flag('home.mainlist.show-logout')
  const { apps, shortcuts } = useApps()
  const { konnectors } = useServices()

  return (
    <div className="app-list-wrapper u-m-auto u-w-100">
      <div className="app-list app-list--gutter u-w-100 u-mh-auto u-flex-justify-center">
        {apps}
        {konnectors}
        {shortcuts &&
          shortcuts.map((shortcut, index) => (
            <ShortcutLink key={index} file={shortcut} />
          ))}
        {<AddTile />}
        {showLogout && <LogoutTile />}
      </div>
    </div>
  )
}

export default ApplicationsAndServices
