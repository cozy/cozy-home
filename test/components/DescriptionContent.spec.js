'use strict'

/* eslint-env jest */

import React from 'react'
import { shallow } from 'enzyme'

import { tMock } from '../jestLib/I18n'
import { DescriptionContent } from 'components/DescriptionContent'

describe('DescriptionContent component', () => {
  beforeEach(() => {
    jest.resetModules()
  })

  it('should be displayed correctly with just title and children', () => {
    const component = shallow(
      <DescriptionContent t={tMock} title="A title mock">
        Test description component children
      </DescriptionContent>
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should handle css from props via cssClassesObject', () => {
    const cssClassesObject = { 'col-mock-class': true }
    const component = shallow(
      <DescriptionContent
        t={tMock}
        title="A title mock"
        cssClassesObject={cssClassesObject}
      />
    ).getElement()
    expect(component).toMatchSnapshot()
  })

  it('should be displayed correctly with two messages', () => {
    const messages = [
      '**First** message `here`',
      'second message with [link](https://examplelink.mock)'
    ]
    const component = shallow(
      <DescriptionContent t={tMock} title="A title mock" messages={messages}>
        Test description component children
      </DescriptionContent>
    ).getElement()
    expect(component).toMatchSnapshot()
  })
})
