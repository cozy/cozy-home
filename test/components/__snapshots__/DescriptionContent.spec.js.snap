// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DescriptionContent component should be displayed correctly with just title and children 1`] = `
<div>
  <div
    class=""
  >
    <h4
      class="col-account-description-title"
    >
      A title mock
    </h4>
    Test description component children
  </div>
</div>
`;

exports[`DescriptionContent component should be displayed correctly with two messages 1`] = `
<div>
  <div
    class=""
  >
    <h4
      class="col-account-description-title"
    >
      A title mock
    </h4>
    <div
      class="col-account-description-message"
    >
      <p>
        <strong>
          First
        </strong>
         message 
        <code>
          here
        </code>
      </p>
    </div>
    <div
      class="col-account-description-message"
    >
      <p>
        second message with 
        <a
          href="https://examplelink.mock"
        >
          link
        </a>
      </p>
    </div>
    Test description component children
  </div>
</div>
`;

exports[`DescriptionContent component should handle css from props via cssClassesObject 1`] = `
<div>
  <div
    class="col-mock-class"
  >
    <h4
      class="col-account-description-title"
    >
      A title mock
    </h4>
  </div>
</div>
`;
